commit f417d92c1714406b85d887c96736801498962ff5
Author: Samuel Thibault <samuel.thibault@ens-lyon.org>
Date:   Sat Jan 6 22:28:36 2018 +0100

    hurd: Fix glob64 compatibility symbols
    
    This follows ccf970c7a ('posix: Add compat glob symbol to not follow
    dangling symbols') by adding to gnu/ the same compatibility as for Linux.
    
            * sysdeps/gnu/glob64.c (__glob): Define macro instead of glob macro.
            (__glob64): Define GLIBC_2_27 versioned symbol instead of glob64.
            * sysdeps/gnu/glob-lstat-compat.c: New file.
            * sysdeps/gnu/glob64-lstat-compat.c: New file.

diff --git a/sysdeps/gnu/glob-lstat-compat.c b/sysdeps/gnu/glob-lstat-compat.c
new file mode 100644
index 0000000000..7cb52d4fee
--- /dev/null
+++ b/sysdeps/gnu/glob-lstat-compat.c
@@ -0,0 +1,38 @@
+/* Compat glob which does not use gl_lstat for GLOB_ALTDIRFUNC.
+   GNU version
+   Copyright (C) 2017-2018 Free Software Foundation, Inc.
+   This file is part of the GNU C Library.
+
+   The GNU C Library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2.1 of the License, or (at your option) any later version.
+
+   The GNU C Library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with the GNU C Library; if not, see
+   <http://www.gnu.org/licenses/>.  */
+
+#include <sys/stat.h>
+#include <shlib-compat.h>
+
+#define glob64 __no_glob64_decl
+#include <glob.h>
+#undef glob64
+
+#define __glob __glob_lstat_compat
+
+#define GLOB_ATTRIBUTE attribute_compat_text_section
+
+/* Avoid calling gl_lstat with GLOB_ALTDIRFUNC.  */
+#define GLOB_NO_LSTAT
+
+#include <posix/glob.c>
+
+#if SHLIB_COMPAT(libc, GLIBC_2_0, GLIBC_2_27)
+compat_symbol (libc, __glob_lstat_compat, glob, GLIBC_2_0);
+#endif
diff --git a/sysdeps/gnu/glob64-lstat-compat.c b/sysdeps/gnu/glob64-lstat-compat.c
new file mode 100644
index 0000000000..e237407e0b
--- /dev/null
+++ b/sysdeps/gnu/glob64-lstat-compat.c
@@ -0,0 +1,48 @@
+/* Compat glob which does not use gl_lstat for GLOB_ALTDIRFUNC.
+   GNU version
+   Copyright (C) 2017-2018 Free Software Foundation, Inc.
+   This file is part of the GNU C Library.
+
+   The GNU C Library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2.1 of the License, or (at your option) any later version.
+
+   The GNU C Library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with the GNU C Library; if not, see
+   <http://www.gnu.org/licenses/>.  */
+
+#include <dirent.h>
+#include <glob.h>
+#include <sys/stat.h>
+#include <shlib-compat.h>
+
+#define dirent dirent64
+#define __readdir(dirp) __readdir64 (dirp)
+
+#define glob_t glob64_t
+#define __glob __glob64_lstat_compat
+#define globfree globfree64
+
+#undef stat
+#define stat stat64
+#undef __stat
+#define __stat(file, buf) __xstat64 (_STAT_VER, file, buf)
+
+#define COMPILE_GLOB64	1
+
+#define GLOB_ATTRIBUTE attribute_compat_text_section
+
+/* Avoid calling gl_lstat with GLOB_ALTDIRFUNC.  */
+#define GLOB_NO_LSTAT
+
+#include <posix/glob.c>
+
+#if SHLIB_COMPAT(libc, GLIBC_2_1, GLIBC_2_27)
+compat_symbol (libc, __glob64_lstat_compat, glob64, GLIBC_2_1);
+#endif
diff --git a/sysdeps/gnu/glob64.c b/sysdeps/gnu/glob64.c
index 52e97e2f6a..42b5b225ca 100644
--- a/sysdeps/gnu/glob64.c
+++ b/sysdeps/gnu/glob64.c
@@ -6,8 +6,7 @@
 #define __readdir(dirp) __readdir64 (dirp)
 
 #define glob_t glob64_t
-#define glob(pattern, flags, errfunc, pglob) \
-  glob64 (pattern, flags, errfunc, pglob)
+#define __glob __glob64
 #define globfree(pglob) globfree64 (pglob)
 
 #undef stat
@@ -19,4 +18,6 @@
 
 #include <posix/glob.c>
 
-libc_hidden_def (glob64)
+libc_hidden_def (__glob64)
+versioned_symbol (libc, __glob64, glob64, GLIBC_2_27);
+libc_hidden_ver (__glob64, glob64)
